tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_shape_calc_placement.cc

```diff
@@ -0,0 +1,539 @@
+  ModuleOp module = getOperation();
+  Builder builder(&getContext());
+
+  module.walk([&](Operation* op) {
+    auto dialect_name = op->getDialect()->getNamespace();
+    if (!isTargetDialect(dialect_name)) {
+      return;
+    }
+    if (isa<mhlo::TupleOp>(op) || isa<mhlo::GetTupleElementOp>(op) ||
+        isa<mhlo::WhileOp>(op) || isa<mhlo::IfOp>(op) ||
+        isa<mhlo::ReturnOp>(op)) {
+      return;
+    }
+    // skip the ops that is already placed on Host
+    auto attr = op->getAttrOfType<StringAttr>(kPlaceTyAttr);
+    if ((attr != nullptr) && (attr.getValue() == kTypeHost)) return;
+
+    if (isa<mhlo::GetDimensionSizeOp>(op) || isa<tensor::FromElementsOp>(op)) {
+      op->setAttr(kPlaceTyAttr, builder.getStringAttr(kTypeHost));
+      return;
+    }
+
+    // ops that only cares about the output element type
+    if (isa<mhlo::ConstOp>(op) || isa<mhlo::SelectOp>(op) ||
+        isa<mhlo::IotaOp>(op) || isa<mhlo::DynamicIotaOp>(op)) {
+      auto result_ty = op->getResult(0).getType().dyn_cast<RankedTensorType>();
+      assert(result_ty && "unexpected non ranked type for ConstOp");
+      auto elem_type = result_ty.getElementType();
+      if (elem_type.isInteger(32)) {
+        op->setAttr(kPlaceTyAttr, builder.getStringAttr(kTypeHost));
+        return;
+      } else {
+        return;
+      }
+    }
+
+    std::string op_name = op->getName().getStringRef().str();
+    bool place_on_host = false;
+    // follow the rule of kPlaceRuleMap exist, or else follow
+    // kShapeCalcOperandMap
+    if (kPlaceRuleMap.find(op_name) != kPlaceRuleMap.end()) {
+      auto dominant_idx = kPlaceRuleMap.at(op_name);
+      auto operand_ty =
+          op->getOperand(dominant_idx).getType().dyn_cast<RankedTensorType>();
+      assert(operand_ty && "unexpected non unranked type of operand");
+      if (operand_ty.getElementType().isInteger(32)) {
+        place_on_host = true;
+      }
+    } else {
+      std::set<int> shape_operand_indices;
+      if (kShapeCalcOperandMap.find(op_name) != kShapeCalcOperandMap.end()) {
+        shape_operand_indices = kShapeCalcOperandMap.at(op_name);
+      }
+      for (int idx = 0; idx < op->getNumOperands(); ++idx) {
+        // if it is not "shape operand", then it must be "data operand"
+        if (shape_operand_indices.find(idx) == shape_operand_indices.end()) {
+          auto operand_ty =
+              op->getOperand(idx).getType().dyn_cast<RankedTensorType>();
+          if (!operand_ty) continue;
+          auto elem_type = operand_ty.getElementType();
+          if (elem_type.isInteger(32)) {
+            place_on_host = true;
+            break;
+          }
+        }
+      }
+    }
+
+    if (!place_on_host) {
+      // For most ops, we can safely omit to insert placement attribute since
+      // currently we suppose ops without placement attribute are placed on
+      // device. However, ops having tuple outputs should have explicit
+      // placement attributes.
```

Comment:
Can you expand the comment with "why"?

Reply:
If elements in tuple placed in different devices, explicit placement attrs for tuple can make downstream pass easier.